numba/tests/test_unicode.py

```diff
@@ -495,6 +507,118 @@ def try_compile_bad_optional(*args):
+
+        starts = list(range(-20, 20)) + [None]
+        ends = list(range(-20, 20)) + [None]
+        for s in UNICODE_EXAMPLES:
+            for sub_str in ['', 'xx', s[:-2], s[3:], s]:
+                for start, end in product(starts, ends):
+                    msg = 'Results of "{}".rfind("{}", {}, {}) must be equal'
+                    self.assertEqual(pyfunc(s, sub_str, start, end),
+                                     cfunc(s, sub_str, start, end),
+                                     msg=msg.format(s, sub_str, start, end))
+
+    def test_rfind_wrong_substr(self):
+        cfunc = njit(rfind_usecase)
+
+        for s in UNICODE_EXAMPLES:
+            for sub_str in [None, 1, False]:
+                with self.assertRaises(TypingError) as raises:
+                    cfunc(s, sub_str)
+                msg = 'must be {}'.format(types.UnicodeType)
+                self.assertIn(msg, str(raises.exception))
+
+    def test_rfind_wrong_start_end(self):
+        cfunc = njit(rfind_with_start_end_usecase)
+
+        for s in UNICODE_EXAMPLES:
+            for sub_str in ['', 'xx', s[:-2], s[3:], s]:
+                for start, end in product([0.1, False], [-0.1, True, None]):
+                    with self.assertRaises(TypingError) as raises:
+                        cfunc(s, sub_str, start, end)
+                    accepted_types = (types.Integer, int, types.NoneType)
+                    msg = 'must be {}'.format(accepted_types)
+                    self.assertIn(msg, str(raises.exception))
+
+    def test_rfind_with_start_end_optional(self):
+        pyfunc = rfind_with_start_end_usecase
+        sig_optional = types.int64(types.unicode_type,
+                                   types.unicode_type,
+                                   types.Optional(types.intp),
+                                   types.Optional(types.intp))
+        cfunc = njit([sig_optional])(pyfunc)
+
+        starts = list(range(-20, 20)) + [None]
+        ends = list(range(-20, 20)) + [None]
+        for s in UNICODE_EXAMPLES:
+            for sub_str in ['', 'xx', s[:-2], s[3:], s]:
+                for start, end in product(starts, ends):
+                    msg = 'Results of "{}".rfind("{}", {}, {}) must be equal'
+                    self.assertEqual(pyfunc(s, sub_str, start, end),
+                                     cfunc(s, sub_str, start, end),
+                                     msg=msg.format(s, sub_str, start, end))
+
+    def test_rfind_wrong_start_end_optional(self):
+        def try_compile_wrong_optional(*args):
+            wrong_sig_optional = types.int64(types.unicode_type,
+                                             types.unicode_type,
+                                             types.Optional(types.float64),
+                                             types.Optional(types.float64))
+            njit([wrong_sig_optional])(rfind_with_start_end_usecase)
+
+        start, end = 0.1, -0.1
+        for s in UNICODE_EXAMPLES:
```

Comment:
I don't think that you need go through all the `UNICODE_EXAMPLES` when testing unacceptable types. Calling the function once with a simple, known to work, `string`  and `sub_string` combination should suffice.

Reply:
Also, I think you need to test start and end being incorrect types individually. For example, a test where start is an integer and end is a float. If both are of unacceptable type, the type-check for `end` will never run.